pkg/api/api.go

```diff
@@ -43,101 +43,124 @@ type podInfo struct {
+		}))
 
-	return a
-}
+	podLister := informerFactory.Core().V1().Pods().Lister()
+
+	ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
+	defer cancel()
 
-// Init handles any api initialization.
-func (a *API) Init() error {
-	a.log.Debugln("Initializing API")
+	informerFactory.Start(ctx.Done())
 
-	a.router = mux.NewRouter()
+	for t, ok := range informerFactory.WaitForCacheSync(ctx.Done()) {
+		if !ok {
+			return nil, fmt.Errorf("timed out while waiting for informer cache to sync: %s", t)
+		}
+	}
 
-	a.router.HandleFunc("/api/configuration/current", a.getCurrentConfiguration)
-	a.router.HandleFunc("/api/status/nodes", a.getMeshNodes)
-	a.router.HandleFunc("/api/status/node/{node}/configuration", a.getMeshNodeConfiguration)
-	a.router.HandleFunc("/api/status/readiness", a.getReadiness)
+	router := mux.NewRouter()
+
+	api := &API{
+		Server: http.Server{
+			Addr:         fmt.Sprintf("%s:%d", apiHost, apiPort),
+			ReadTimeout:  5 * time.Second,
+			WriteTimeout: 5 * time.Second,
+			Handler:      router,
+		},
+		configuration: safe.New(provider.NewDefaultDynamicConfig()),
+		topology:      safe.New(topology.NewTopology()),
+		readiness:     safe.New(false),
+		podLister:     podLister,
+		namespace:     namespace,
+		log:           log,
+	}
 
-	return nil
+	router.HandleFunc("/api/configuration/current", api.getCurrentConfiguration)
+	router.HandleFunc("/api/topology/current", api.getCurrentTopology)
+	router.HandleFunc("/api/status/nodes", api.getMeshNodes)
+	router.HandleFunc("/api/status/node/{node}/configuration", api.getMeshNodeConfiguration)
+	router.HandleFunc("/api/status/readiness", api.getReadiness)
+
+	return api, nil
 }
 
-// Start runs the API.
-func (a *API) Start() {
-	a.log.Debug("Starting API")
+// SetReadiness sets the readiness flag in the API.
+func (a *API) SetReadiness(isReady bool) {
+	a.readiness.Set(isReady)
+	a.log.Debugf("API readiness: %s", isReady)
+}
 
-	go a.Run()
+// SetConfig sets the current dynamic configuration.
+func (a *API) SetConfig(cfg *dynamic.Configuration) {
+	a.configuration.Set(cfg)
 }
 
-// Run wraps the listenAndServe method.
-func (a *API) Run() {
-	a.log.Error(http.ListenAndServe(fmt.Sprintf("%s:%d", a.apiHost, a.apiPort), a.router))
+// SetTopology sets the current topology.
+func (a *API) SetTopology(topo *topology.Topology) {
+	a.topology.Set(topo)
 }
 
-// EnableReadiness enables the readiness flag in the API.
-func (a *API) EnableReadiness() {
-	if !a.readiness {
-		a.log.Debug("API readiness enabled")
+// getCurrentConfiguration returns the current configuration.
+func (a *API) getCurrentConfiguration(w http.ResponseWriter, _ *http.Request) {
+	w.Header().Set("Content-Type", "application/json")
 
-		a.readiness = true
+	if err := json.NewEncoder(w).Encode(a.configuration.Get()); err != nil {
+		a.log.Errorf("Unable to serialize dynamic configuration: %v", err)
+		a.writeErrorResponse(w, nil, http.StatusInternalServerError)
```

Comment:
Why not using https://golang.org/pkg/net/http/#Error?
```suggestion
		http.Error(w, "", http.StatusInternalServerError)
```

Reply:
I just followed what was already present in the code ;) Yeah It's definitely more appropriate.